src/state-receiver/connection.js

```diff
@@ -1,251 +1,298 @@
+    if (code !== 1000) {
+      // 1000 = closed by me normally
+      this.reconnect();
     }
+  }
 
-    get_blocks_result_v0(response) {
-        this.blocksQueue.push(response);
-        this.processBlocks();
-    }
+  onOpen() {
+    this.requestBlocks(this.currentArgs);
+  }
 
-    async processBlocks() {
-        if (this.inProcessBlocks)
-            return;
-        this.inProcessBlocks = true;
-        while (this.blocksQueue.length) {
-            let response = this.blocksQueue.shift();
-            if (response.this_block){
-                let block_num = response.this_block.block_num;
-                this.currentArgs.start_block_num = block_num - 50; // replay 25 seconds
-            }
-            this.send(['get_blocks_ack_request_v0', { num_messages: 1 }]);
-            let block, traces = [], deltas = [];
-            if (this.currentArgs.fetch_block && response.block && response.block.length)
-                block = this.deserialize('signed_block', response.block);
-            if (this.currentArgs.fetch_traces && response.traces && response.traces.length)
-                traces = this.deserialize('transaction_trace[]', response.traces);
-            if (this.currentArgs.fetch_deltas && response.deltas && response.deltas.length)
-                deltas = this.deserialize('table_delta[]', response.deltas);
-            await this.receivedBlock(response, block, traces, deltas);
-        }
-        this.inProcessBlocks = false;
+  requestStatus() {
+    this.send(["get_status_request_v0", {}]);
+  }
+
+  requestBlocks(requestArgs) {
+    if (!this.currentArgs) {
+      this.currentArgs = {
+        start_block_num: 0,
+        end_block_num: 0xffffffff,
+        max_messages_in_flight: 5,
+        have_positions: [],
+        irreversible_only: false,
+        fetch_block: false,
+        fetch_traces: false,
+        fetch_deltas: false,
+        ...requestArgs,
+      };
     }
+    this.send(["get_blocks_request_v0", this.currentArgs]);
+  }
 
-    forEachRow(delta, f) {
-        const type = this.tables.get(delta.name);
-        for (let row of delta.rows) {
-            let data;
-            try {
-                data = this.deserialize(type, row.data);
-            } catch (e) {
-                console.error(e);
-            }
-            if (data)
-                f(row.present, data[1]);
-        }
+  get_status_result_v0(response) {
+    this.logger.info(response);
+  }
+
+  get_blocks_result_v0(response) {
+    this.blocksQueue.push(response);
+    this.processBlocks();
+  }
+
+  async processBlocks() {
+    if (this.inProcessBlocks) return;
+    this.inProcessBlocks = true;
+    while (this.blocksQueue.length) {
+      let response = this.blocksQueue.shift();
+      if (response.this_block) {
+        let block_num = response.this_block.block_num;
+        this.currentArgs.start_block_num = block_num - 50; // replay 25 seconds
+      } else {
+        this.logger.info("Stop processing blocks");
+        break;
```

Comment:
This will solve the WS error but are we sure that it won't happen that during processing the range we will receive a message without this_block and it won't be at the end?
I would set some flag that we are done with processing eg. it in onComplete where we call done_handlers. Anyway we can think about it after some tests and in the proper refactor